PhotonSieveDevelopment.js

```diff
@@ -24,92 +24,96 @@ class Sieve {
-    
-    // loop through all pinholes on current ring, check if min spacing constraint is violated
-
-    for(i = 0; i < currentPinholeArray.length; i ++){
-        let conflictDistanceBetweenCenters = distanceBetween(candidateX, currentPinholeArray[i][1], candidateY, currentPinholeArray[i][2]);
-        let conflictDistanceBetweenPerimeters = conflictDistanceBetweenCenters - currentPinholeArray[i][0] - conflictRadius;
-        if(conflictDistanceBetweenPerimeters <= SPACE_MIN){
-            conflictArray.push(i);
-        }
-    return conflictArray.length
-    }
+function distanceBetween(a, b) {
+    return Math.hypot(b.x - a.x, b.y - a.y);
 }
 
 function randMinMax(min, max) {
     return Math.random() * ((max - min) + min);
- }
-
-let currentPinholeArray = []; 
+ } 
  
 //create new pinhole sieve object
 const photonSieve = new Sieve(0, 0, 0);
 
 //loop through fresnelZonePlate, add a Ring for each even-numbered zone
-for(i = 1; i < fresnelZonePlate.length; i = i + 2){
-    let zone = i - 1;
+for(let i = 1; i < fresnelZonePlate.length; i = i + 2){
     let zoneWidth = fresnelZonePlate[i] - fresnelZonePlate[i-1];
     let centerlineRadius = fresnelZonePlate[i-1] + zoneWidth / 2;
     let ringArea = (Math.PI * fresnelZonePlate[i] ** 2) - (Math.PI * fresnelZonePlate[i-1] ** 2);
-    var ring = new Ring(i-1, centerlineRadius, zoneWidth, 0, 0);
+    var ring = new Ring(i-1, centerlineRadius, zoneWidth);
 
     //add pinholes to ring until total area constraint is fulfilled
     while(ring.openArea < AREA_PROP * ringArea){
         let minRadius = Math.max(DIA_MIN, zoneWidth * .25);
         let candidateRadius = randMinMax(minRadius, zoneWidth * MAX_FACTOR);
-        let candidateAzimuth = Math.random() * Math.PI * 2;
-        if(getConflicts(candidateAzimuth, candidateRadius) = 0){
-            let pinhole = new Pinhole(candidateRadius, candidateAzimuth, 0);         
-            currentPinholeArray.push([pinhole.radius, pinhole.pointX, pinhole.pointY]);
-            pinhole.pinholeArea = Math.PI * radius ** 2;
-            ring.openArea = ring.openArea + pinhole.pinholeArea;
-            ring.numOfPinholes = ring.numOfPinholes + 1;
-            }
+        let candidateAzimuth = Math.random() * Math.PI * 2 * ring.centerlineRadius;  
+        if(ring.getConflicts(candidateRadius, candidateAzimuth).length == 0){
+            let pinhole = new Pinhole(candidateRadius, candidateAzimuth, ring.centerlineRadius);         
+            ring.addPinhole(pinhole);
         }
-    photonSieve.totalArea = photonSieve.totalArea + ring.openArea;
-    photonSieve.totalPinholes = photonSieve.totalPinholes + ring.numOfPinholes;
-    let sieveEquivalentAperture = Math.sqrt(photonSieve.totalArea / Math.PI) * 2;
-    photonSieve.fNumber = FD / sieveEquivalentAperture;
-    console.lot(photonSieve);
     }
-console.log(photonSieve);
\ No newline at end of file
+
+    photonSieve.addRing(ring);
+}
+console.log(JSON.stringify(photonSieve, null, 2));
```

Comment:
Added JSON.stringify with pretty print (the "2" flag) to make the output to console very readable 

Sample output:

```
{
  "totalPinholes": 484,
  "totalArea": 0.7186523407031569,
  "fNumber": 75.26932387427817,
  "rings": [
    {
      "zone": 0,
      "centerlineRadius": 0.23187764063355398,
      "openArea": 0.04964989107381159,
      "pinholes": [
        {
          "radius": 0.1208282965917095,
          "azimuth": 1.375305197811141,
          "pinholeArea": 0.045865610497640805,
          "point": {
            "x": 0.04504184504151386,
            "y": 0.2274609250421704
          }
        },
        {
          "radius": 0.034706972202836306,
          "azimuth": 0.8411142387845659,
          "pinholeArea": 0.0037842805761707835,
          "point": {
            "x": 0.15457721797703616,
            "y": 0.17283843296056425
          }
        }
      ]
    },
    {
      "zone": 2,
      "centerlineRadius": 0.3584517747191267,
      "openArea": 0.05275838720997238,
      "pinholes": [
        {
          "radius": 0.045126649785063554,
          "azimuth": 0.4369649015187486,
          "pinholeArea": 0.006397584898283555,
          "point": {
            "x": 0.3247717399726112,
            "y": 0.15169374316185277
          }
        },
        {
          "radius": 0.026427751595375557,
          "azimuth": 0.45289767195310715,
          "pinholeArea": 0.002194170161537513,
          "point": {
            "x": 0.32231371934510517,
            "y": 0.15684878425163623
          }
        },
        {
          "radius": 0.061054314011740565,
          "azimuth": 1.7103101627951836,
          "pinholeArea": 0.011710692696776322,
          "point": {
            "x": -0.04984690986378331,
            "y": 0.35496895691922614
          }
        },
        {
          "radius": 0.057003298018205556,
          "azimuth": 1.6726114685412867,
          "pinholeArea": 0.010208215723077424,
          "point": {
            "x": -0.03643279635251278,
            "y": 0.3565954656879808
          }
        },
        {
          "radius": 0.05615826706303603,
          "azimuth": 0.9653815782733765,
          "pinholeArea": 0.00990780084569009,
          "point": {
            "x": 0.20399603729158391,
            "y": 0.294742754904378
          }
        },
        {
          "radius": 0.06267311583858072,
          "azimuth": 1.2672069289344272,
          "pinholeArea": 0.01233992288460747,
          "point": {
            "x": 0.10715821831432258,
            "y": 0.342059630835022
          }
        }
      ]
    },
    {
      "zone": 4,
      "centerlineRadius": 0.45003312306951715,
      "openArea": 0.04727237783962021,
      "pinholes": [
        {
          "radius": 0.014912629690290682,
          "azimuth": 0.46841452730205435,
          "pinholeArea": 0.0006986478709345967,
          "point": {
            "x": 0.4015578979957083,
            "y": 0.2031774259532084
          }
        },
        {
          "radius": 0.006305650374250137,
          "azimuth": 0.46702392392747505,
          "pinholeArea": 0.00012491357751710842,
          "point": {
            "x": 0.40184004885701513,
            "y": 0.20261882191517863
          }
        },
        {
          "radius": 0.03372950374940499,
          "azimuth": 2.072165186409643,
          "pinholeArea": 0.0035741253180060968,
          "point": {
            "x": -0.21629778919570392,
            "y": 0.3946455096016601
          }
        },
        {
          "radius": 0.012178961387506832,
          "azimuth": 0.1608003480805312,
          "pinholeArea": 0.0004659833291911611,
          "point": {
            "x": 0.44422745149525605,
            "y": 0.07205402971196742
          }
        },
        {
          "radius": 0.05983926583484357,
          "azimuth": 0.7638729587395042,
          "pinholeArea": 0.011249219364759458,
          "point": {
            "x": 0.3249970078343427,
            "y": 0.3112985010539359
          }
        },
        {
          "radius": 0.015379380673219992,
          "azimuth": 1.3944915172478318,
          "pinholeArea": 0.0007430663016078739,
          "point": {
            "x": 0.07893260087302607,
            "y": 0.4430569448492177
          }
        },
        {
          "radius": 0.038329137467237435,
          "azimuth": 2.362217814222325,
          "pinholeArea": 0.00461538532967248,
          "point": {
            "x": -0.3201324399477447,
            "y": 0.3162989610365592
          }
        },
        {
          "radius": 0.012023254424032013,
          "azimuth": 1.1535531782834854,
          "pinholeArea": 0.0004541443832553095,
          "point": {
            "x": 0.18237216809624093,
            "y": 0.4114246032550552
          }
        },
        {
          "radius": 0.039132745671554045,
          "azimuth": 0.8085449723247334,
          "pinholeArea": 0.004810946345883598,
          "point": {
            "x": 0.31077107545423976,
            "y": 0.32550138328541467
          }
        },
        {
          "radius": 0.02292687013581751,
          "azimuth": 0.5422655054522454,
          "pinholeArea": 0.0016513510796869746,
          "point": {
            "x": 0.38547213780577627,
            "y": 0.23225211050741343
          }
        },
        {
          "radius": 0.04132556714030324,
          "azimuth": 2.7327539872035267,
          "pinholeArea": 0.005365219786110048,
          "point": {
            "x": -0.4129427983690217,
            "y": 0.17890795715916222
          }
        },
        {
          "radius": 0.027733356283791408,
          "azimuth": 1.2034086132283,
          "pinholeArea": 0.0024163215914683057,
          "point": {
            "x": 0.16164230976991514,
            "y": 0.4200018756528951
          }
        },
        {
          "radius": 0.0011019597913367238,
          "azimuth": 1.2103309462491776,
          "pinholeArea": 0.000003814884282361672,
          "point": {
            "x": 0.15873106730818082,
            "y": 0.42111074568444457
          }
        },
        {
          "radius": 0.015647669836092903,
          "azimuth": 2.024737128755343,
          "pinholeArea": 0.0007692176144287163,
          "point": {
            "x": -0.1973443088343743,
            "y": 0.4044564693887171
          }
        },
        {
          "radius": 0.009379454596431787,
          "azimuth": 1.2202364050070047,
          "pinholeArea": 0.0002763790015486025,
          "point": {
            "x": 0.15455205326451535,
            "y": 0.42266236488647246
          }
        },
        {
          "radius": 0.05049324600587756,
          "azimuth": 0.7882982469964981,
          "pinholeArea": 0.00800970375999557,
          "point": {
            "x": 0.3172972673021903,
            "y": 0.3191430024648286
          }
        },
        {
          "radius": 0.017197531574585128,
          "azimuth": 0.4889760215201528,
          "pinholeArea": 0.0009291420251021824,
          "point": {
            "x": 0.397295679548335,
            "y": 0.21139052692098076
          }
        },
        {
          "radius": 0.018837480611422305,
          "azimuth": 1.3378083170583766,
          "pinholeArea": 0.001114796276169764,
          "point": {
            "x": 0.10390626941947784,
            "y": 0.4378736108000002
          }
        }
      ]
    },
    {
      "zone": 6,
      "centerlineRadius": 0.5257776693355838,
      "openArea": 0.046376265067083824,
      "pinholes": [
        {
          "radius": 0.035634963497511093,
          "azimuth": 3.045222367282718,
          "pinholeArea": 0.003989353389846465,
          "point": {
            "x": -0.5233380487033584,
            "y": 0.050590951279054656
          }
        },
        {
          "radius": 0.03801802202064353,
          "azimuth": 1.531875300687872,
          "pinholeArea": 0.004540763768573579,
          "point": {
            "x": 0.020458640199207092,
            "y": 0.5253794834337917
          }
        },
        {
          "radius": 0.04003654355249743,
          "azimuth": 1.0445559640951585,
          "pinholeArea": 0.0050357368376095675,
          "point": {
            "x": 0.2640907279188884,
            "y": 0.45464078677482306
          }
        },
        {
          "radius": 0.0514951721220229,
          "azimuth": 1.6814379150331276,
          "pinholeArea": 0.00833072696443256,
          "point": {
            "x": -0.058054261217559064,
            "y": 0.522562781229626
          }
        },
        {
          "radius": 0.028096534773612625,
          "azimuth": 2.3521552517906987,
          "pinholeArea": 0.0024800212011920316,
          "point": {
            "x": -0.37027621467308874,
            "y": 0.3732796303300345
          }
        },
        {
          "radius": 0.02290930888609563,
          "azimuth": 2.1325088694443424,
          "pinholeArea": 0.001648822284255105,
          "point": {
            "x": -0.28004831384479045,
            "y": 0.4449888756864021
          }
        },
        {
          "radius": 0.015560826755469475,
          "azimuth": 1.491887262169545,
          "pinholeArea": 0.0007607031381171885,
          "point": {
            "x": 0.04144558169979302,
            "y": 0.5241416042726662
          }
        },
        {
          "radius": 0.014294276441114731,
          "azimuth": 1.019317108806252,
          "pinholeArea": 0.0006419101254587818,
          "point": {
            "x": 0.2754800143547785,
            "y": 0.4478313513623734
          }
        },
        {
          "radius": 0.05311631176133847,
          "azimuth": 1.4184702896966408,
          "pinholeArea": 0.008863509107281302,
          "point": {
            "x": 0.07978026499426266,
            "y": 0.5196895870511585
          }
        },
        {
          "radius": 0.014804857221954149,
          "azimuth": 2.4069313059864372,
          "pinholeArea": 0.0006885862075797802,
          "point": {
            "x": -0.3901574241751584,
            "y": 0.352447644243743
          }
        },
        {
          "radius": 0.02682716746600672,
          "azimuth": 1.9341840269873856,
          "pinholeArea": 0.0022609945386164337,
          "point": {
            "x": -0.1868838573240624,
            "y": 0.49144336544879513
          }
        },
        {
          "radius": 0.006513081324499825,
          "azimuth": 1.8822440962212663,
          "pinholeArea": 0.0001332670777151268,
          "point": {
            "x": -0.16111777144958955,
            "y": 0.5004829880176511
          }
        },
        {
          "radius": 0.01636945117212559,
          "azimuth": 2.138855977746614,
          "pinholeArea": 0.0008418178112189882,
          "point": {
            "x": -0.2828670465043297,
            "y": 0.4432024273104513
          }
        },
        {
          "radius": 0.041719751556822976,
          "azimuth": 0.893626241403512,
          "pinholeArea": 0.00546806035725216,
          "point": {
            "x": 0.3294470533537792,
            "y": 0.40976431836907257
          }
        },
        {
          "radius": 0.0148414277232104,
          "azimuth": 2.6970244933265852,
          "pinholeArea": 0.0006919922579347614,
          "point": {
            "x": -0.4746702239052612,
            "y": 0.22612018067763806
          }
        }
      ]
    },
    {
      "zone": 8,
      "centerlineRadius": 0.591867441654227,
      "openArea": 0.0476163011061214,
      "pinholes": [
        {
          "radius": 0.018081175028404034,
          "azimuth": 2.3759696422700096,
          "pinholeArea": 0.0010270774003513492,
          "point": {
            "x": -0.42670728133254365,
            "y": 0.41015602463953793
          }
        },
        {
          "radius": 0.008554633673812945,
          "azimuth": 2.283815115385096,
          "pinholeArea": 0.0002299072710889022,
          "point": {
            "x": -0.3871524088346753,
            "y": 0.44768301377629705
          }
        },
        {
          "radius": 0.00555773231898395,
          "azimuth": 3.056632076556619,
          "pinholeArea": 0.00009703873448543755,
          "point": {
            "x": -0.5897325880121966,
            "y": 0.05022492535341992
          }
        },
        {
          "radius": 0.04721904255162753,
          "azimuth": 3.1801670840727154,
          "pinholeArea": 0.007004614296538149,
          "point": {
            "x": -0.5914271508170151,
            "y": -0.02282528787961786
          }
        },
        {
          "radius": 0.01795762954546554,
          "azimuth": 0.7570119597323632,
          "pinholeArea": 0.001013089674211314,
          "point": {
            "x": 0.4302232922521752,
            "y": 0.4064664651530545
          }
        },
        {
          "radius": 0.023797526184381187,
          "azimuth": 0.0676135208409321,
          "pinholeArea": 0.0017791538280068258,
          "point": {
            "x": 0.5905150698729885,
            "y": 0.03998775741673024
          }
        },
        {
          "radius": 0.0077583046017496065,
          "azimuth": 3.0539680402257297,
          "pinholeArea": 0.0001890965153962417,
          "point": {
            "x": -0.5895966944533424,
            "y": 0.0517958143097665
          }
        },
        {
          "radius": 0.007157190209494503,
          "azimuth": 0.03104547530745047,
          "pinholeArea": 0.00016092925139405402,
          "point": {
            "x": 0.5915822372688313,
            "y": 0.018371854515100233
          }
        },
        {
          "radius": 0.02402757810673382,
          "azimuth": 2.7236264640837726,
          "pinholeArea": 0.0018137184383329203,
          "point": {
            "x": -0.5409173371605831,
            "y": 0.24024051042533157
          }
        },
        {
          "radius": 0.04666379368173291,
          "azimuth": 1.7476325388404743,
          "pinholeArea": 0.006840848290568178,
          "point": {
            "x": -0.10411895866156659,
            "y": 0.5826373751636182
          }
        },
        {
          "radius": 0.04284356546706986,
          "azimuth": 1.919023725656838,
          "pinholeArea": 0.005766616688968469,
          "point": {
            "x": -0.20196419162602386,
            "y": 0.5563430001277687
          }
        },
        {
          "radius": 0.011860398522949796,
          "azimuth": 2.748283225617629,
          "pinholeArea": 0.00044192486387924505,
          "point": {
            "x": -0.5466758713351436,
            "y": 0.2268315678874555
          }
        },
        {
          "radius": 0.03168713126119564,
          "azimuth": 2.8136896944099976,
          "pinholeArea": 0.0031543924054702284,
          "point": {
            "x": -0.5603326226397911,
            "y": 0.1906158978048087
          }
        },
        {
          "radius": 0.04485976947655086,
          "azimuth": 0.1873130795597854,
          "pinholeArea": 0.006322137655276388,
          "point": {
            "x": 0.5815145783525975,
            "y": 0.1102173473357102
          }
        },
        {
          "radius": 0.010457641705541844,
          "azimuth": 2.9668881511669527,
          "pinholeArea": 0.0003435717041422422,
          "point": {
            "x": -0.5828580025758326,
            "y": 0.10287670933515745
          }
        },
        {
          "radius": 0.01767661712384467,
          "azimuth": 0.4374716987727987,
          "pinholeArea": 0.0009816308148298731,
          "point": {
            "x": 0.5361287240308904,
            "y": 0.2507450094405254
          }
        },
        {
          "radius": 0.04141522826280003,
          "azimuth": 2.5739773595758293,
          "pinholeArea": 0.005388526107761132,
          "point": {
            "x": -0.4990540114914762,
            "y": 0.3182014489353958
          }
        },
        {
          "radius": 0.01101508590287034,
          "azimuth": 0.3969853355669738,
          "pinholeArea": 0.00038117608881791466,
          "point": {
            "x": 0.5458383679106755,
            "y": 0.2288395608432463
          }
        },
        {
          "radius": 0.026343450108518685,
          "azimuth": 0.4592341266310326,
          "pinholeArea": 0.002180194187306246,
          "point": {
            "x": 0.5305453834229726,
            "y": 0.2623521766993573
          }
        },
        {
          "radius": 0.007077096823584129,
          "azimuth": 2.164585269504185,
          "pinholeArea": 0.00015734760880617305,
          "point": {
            "x": -0.33115303467847235,
            "y": 0.4905555382559229
          }
        },
        {
          "radius": 0.008771365557188363,
          "azimuth": 1.5488028946823822,
          "pinholeArea": 0.00024170425449308015,
          "point": {
            "x": 0.013016146996041048,
            "y": 0.5917243010116259
          }
        },
        {
          "radius": 0.004070887395768632,
          "azimuth": 1.2424377320367357,
          "pinholeArea": 0.000052062863606627803,
          "point": {
            "x": 0.1908711854882782,
            "y": 0.56024571309437
          }


......ETC ETC

```